### PR TITLE
Feat/26

### PR DESCRIPTION
**Pull request by geneaky**
#26 

카카오 책 검색 api를 호출하는 controller servicce test 작성을 통한 기능 구현

### CODE REVIEW

File: back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookDto.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+@Getter
+@Builder
+public class BookDto {
+
+    private String isbn;
+    private String title;
+    private List<String> author;
+    private String bookCoverImageUrl;
+
```
Comment (Kova700) [inline back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookDto.java:15]: I think publishers should also be added to the DTO.

File: back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.util.StringUtils;
+
+@Getter
+@Setter
+@NoArgsConstructor
+public class BookSearchRequestDto {
+
+    private String isbn;
+    private String title;
+    private String author;
```
Comment (Kova700) [inline back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java:16]: I think publishers should also be added to the DTO.

File: back/src/main/java/toy/bookchat/bookchat/domain/book/dto/KakaoBook.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    private List<Document> documents;
+    private Meta meta;
+
+    public List<BookDto> getBookDtos() {
+        List<BookDto> list = new ArrayList<>();
+
+        for (Document document : documents) {
+            list.add(
+                BookDto.builder()
+                    .isbn(document.getIsbn())
```
Comment (Kova700) [inline back/src/main/java/toy/bookchat/bookchat/domain/book/dto/KakaoBook.java:23]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        );
+        UserPrincipal userPrincipal = new UserPrincipal(1L, "test@gmail.com", "password",
+            "testUser", "somethingImageUrl.com", authorities);
+
+        return userPrincipal;
+    }
+
+    private BookDto getBookDto(String isbn, String title, List<String> author) {
+        BookDto bookDto = BookDto.builder()
+            .isbn(isbn)
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java:79]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    public void 로그인하지_않은_사용자_요청_401() throws Exception {
+        mockMvc.perform(get("/v1/api/books")
+                .param("isbn", "234134"))
+            .andExpect(status().isUnauthorized());
+    }
+
+    @Test
+    public void 로그인한_사용자_요청_200() throws Exception {
+        List<BookDto> bookDtos = new ArrayList<>();
+        bookDtos.add(getBookDto("213123", "effectiveJava", List.of("Joshua")));
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java:97]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        mockMvc.perform(get("/v1/api/books")
+                .param("isbn", "213123")
+                .with(user(getUserPrincipal())))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void 사용자가_isbn으로_책_검색_요청시_성공() throws Exception {
+        List<BookDto> bookDtos = new ArrayList<>();
+        bookDtos.add(getBookDto("213123", "effectiveJava", List.of("Joshua")));
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java:109]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        mockMvc.perform(get("/v1/api/books")
+                .param("isbn", "123456")
+                .with(user(getUserPrincipal())))
+            .andExpect(status().isNotFound());
+    }
+
+    @Test
+    public void 사용자가_도서명_검색_요청시_성공() throws Exception {
+        List<BookDto> bookDtos = new ArrayList<>();
+        bookDtos.add(getBookDto("213123", "effectiveJava", List.of("Joshua")));
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java:146]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        mockMvc.perform(get("/v1/api/books")
+                .param("title", "effectiveJava")
+                .with(user(getUserPrincipal())))
+            .andExpect(status().isNotFound());
+    }
+
+    @Test
+    public void 사용자가_작가명_검색_요청시_성공() throws Exception {
+        List<BookDto> bookDtos = new ArrayList<>();
+        bookDtos.add(getBookDto("213123", "effectiveJava", List.of("Joshua")));
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/api/BookControllerTest.java:184]: I think publishers should also be added to the DTO.

File: back/src/test/java/toy/bookchat/bookchat/domain/book/service/BookSearchServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Autowired
+    private BookSearchService bookSearchService;
+    @Autowired
+    private MockRestServiceServer mockServer;
+
+    @Value("${book.api.uri}")
+    private String apiUri;
+
+    private String isbn = "8996991341 9788996991342";
```
Comment (Kova700) [inline back/src/test/java/toy/bookchat/bookchat/domain/book/service/BookSearchServiceTest.java:29]: I think publishers should also be added to the DTO.

**Review by Kova700 — CHANGES_REQUESTED** [review]
publisher is not exist in BookDTO

File: back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java (hunk 2 above)
Comment (geneaky) [reply back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java:16]: 이부분은 검색 조건을 받는 dto로 출판사는 따로 추가하지 않았습니다

**Review by Kova700 — APPROVED** [review]
Good!

File: back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java (hunk 2 above)
Comment (Kova700) [reply back/src/main/java/toy/bookchat/bookchat/domain/book/dto/BookSearchRequestDto.java:16]: 네 확인했습니다.

**Pull request merged**